pkg/mqtt/subscription_manager.go

```diff
@@ -0,0 +1,199 @@
+package mqtt
+
+import (
+	"sync"
+)
+
+type OnClientConnectFunc func(clientID string)
+type OnClientDisconnectFunc func(clientID string)
+type OnTopicSubscribeFunc func(clientID string, topic string)
+type OnTopicUnsubscribeFunc func(clientID string, topic string)
+
+// SubscriptionManager keeps track of subscribed topics of clients of the
+// mqtt broker by subscribing to broker events.
+// This allows to get notified when a client connects or disconnects
+// or a topic is subscribed or unsubscribed.
+type SubscriptionManager struct {
+	// subscribers keeps track of the clients and their
+	// subscribed topics (and the count of subscriptions per topic).
+	subscribers     *ShrinkingMap[string, *ShrinkingMap[string, int]]
```

Comment:
The topic name should not be duplicated in a subscriber. here we take the advantage of map's feature to avoid iterating topic items. maybe use `bool` for reducing memory?
```suggestion
	subscribers     *ShrinkingMap[string, *ShrinkingMap[string, bool]]
```

Reply:
Normally you use `struct{}` in go to reduce memory instead of bool.
I would leave it as it is (see other comment).